src/clustering/administration/main/command_line.cc

```diff
@@ -645,6 +650,244 @@ service_address_ports_t get_service_address_ports(const std::map<std::string, op
+    'Curve25519' (though not currently supported by OpenSSL) would be other
+    good choices.
+    NOTE: For compatibility reasons, we can only support a single elliptic
+    curve since OpenSSL is ridiculous. */
+    boost::optional<std::string> curve_opt = get_optional_option(opts, "--tls-ecdh-curve");
+    std::string curve_name = curve_opt ? *curve_opt : "prime256v1";
+
+    int curve_nid = OBJ_txt2nid(curve_name.c_str());
+    if (NID_undef == curve_nid) {
+        logNTC("No elliptic curve found corresponding to name: %s", curve_name.c_str());
+        return false;
+    }
+
+    EC_KEY *ec_key = EC_KEY_new_by_curve_name(curve_nid);
+    if (NULL == ec_key) {
+        logNTC("Unable to get Elliptic Curve by name: %s", curve_name.c_str());
+        return false;
+    }
+
+    SSL_CTX_set_tmp_ecdh(*tls_ctx, ec_key);
+    EC_KEY_free(ec_key);
+
+
+    /* If we only supported OpenSSL 1.0.2 then we could have multiple ECDHE
+    curves with something like this:
+
+    if (0 == SSL_CTX_set1_curves_list(*tls_ctx, curves.c_str())) {
+        logNTC("Not able to set ECDHE curves for perfect forward secrecy\n");
+        return false;
+    }
+
+    // Enabling this option allows the server to select the appropriate curve
+    // that is shared with the client.
+    if (0 == SSL_CTX_set_ecdh_auto(*tls_ctx, 1)) {
+        logNTC("Not able to set ECDHE curve selection mode\n");
+        return false;
+    }
+
+    */
+
+    /* If the client and server do not support ECDHE but do support DHE, an
+    admin must specify a file containing parameters for DHE. */
+    boost::optional<std::string> dhparams_filename = get_optional_option(opts, "--tls-dhparams");
+    if (dhparams_filename) {
+        fp_wrapper_t dhparams_fp(dhparams_filename->c_str(), "r");
+        if (NULL == dhparams_fp.get()) {
+            logNTC("unable to open %s for reading: %s", dhparams_filename->c_str(), errno_string(get_errno()).c_str());
+            return false;
+        }
+
+        DH *dhparams = PEM_read_DHparams(dhparams_fp.get(), NULL, NULL, NULL);
+        if (NULL == dhparams) {
+            logNTC("unable to read DH parameters from %s: %s", dhparams_filename->c_str(), ERR_error_string(ERR_get_error(), NULL));
+            return false;
+        }
+
+        if (1 != SSL_CTX_set_tmp_dh(*tls_ctx, dhparams)) {
+            logNTC("unable to set DH parameters: %s", ERR_error_string(ERR_get_error(), NULL));
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool configure_tls(const std::map<std::string, options::values_t> &opts, tls_configs_t *tls_configs) {
+    // Setup OpenSSL context.
+    SSL_library_init();
+    SSL_load_error_strings();
+
+    logNTC("%s\n", SSLeay_version(SSLEAY_VERSION));
+
+    if(!exists_option(opts, "--no-http-admin") && exists_option(opts, "--web-tls")) {
+        if (!(initialize_tls_ctx(opts, &(tls_configs->web)) && configure_web_tls(opts, tls_configs->web))) {
+            SSL_CTX_free(tls_configs->web);
```

Comment:
Once we have a wrapper around SSL_ctx, this would probably become something like `tls_configs->web.reset()`;

(Alternatively, and somewhat cleaner IMO, `initialize_tls_ctx` should take care of resetting the SSL context itself before returning `false`.)